beefy-gadget/src/worker.rs

```diff
@@ -334,26 +283,59 @@ where
 		Ok(sig)
 	}
 
+	/// Return the current active validator set.
+	///
+	/// Note that the validator set could be `None`. This is the case if we don't find
+	/// a BEEFY authority set change and we can't fetch the validator set from the
+	/// BEEFY on-chain state. Such a failure is usually an indication that the BEEFT
+	/// pallet has not been deplyed (yet).
+	fn validator_set(&self, header: &B::Header) -> Option<ValidatorSet<P::Public>> {
+		if let Some(new) = find_authorities_change::<B, P::Public>(header) {
+			Some(new)
+		} else {
+			let at = BlockId::hash(self.client.info().finalized_hash);
+			self.client.runtime_api().validator_set(&at).ok()
+		}
+	}
+
+	/// Return the local authority id.
+	///
+	/// `None` is returned, if we are not permitted to vote
+	fn local_id(&self) -> Option<P::Public> {
+		self.rounds
+			.validators()
+			.iter()
+			.find(|id| SyncCryptoStore::has_keys(&*self.key_store, &[(id.to_raw_vec(), KEY_TYPE)]))
+			.cloned()
+	}
+
 	fn handle_finality_notification(&mut self, notification: FinalityNotification<B>) {
 		debug!(target: "beefy", "🥩 Finality notification: {:?}", notification);
 
-		if let Some(new) = find_authorities_change::<B, P::Public>(&notification.header) {
-			debug!(target: "beefy", "🥩 New validator set: {:?}", new);
+		if let Some(active) = self.validator_set(&notification.header) {
+			debug!(target: "beefy", "🥩 Active validator set id: {:?}", active);
 
 			if let Some(metrics) = self.metrics.as_ref() {
-				metrics.beefy_validator_set_id.set(new.id);
+				metrics.beefy_validator_set_id.set(active.id);
 			}
 
-			self.rounds = round::Rounds::new(new);
-
-			// NOTE: currently we act as if this block has been finalized by BEEFY as we perform
-			// the validator set changes instantly (insecure). Once proper validator set changes
-			// are implemented this should be removed
-			self.best_finalized_block = *notification.header.number();
+			// Authority set change or genesis set id triggers new voting rounds
+			//
+			// TODO: (adoerr) this is still a bit crude and needs to be replaced with
+			// a proper round live cycle handling.
+			if (active.id != self.rounds.validator_set_id()) || (active.id == GENESIS_AUTHORITY_SET_ID) {
+				self.rounds = round::Rounds::new(active.clone());
+				debug!(target: "beefy", "🥩 New Rounds for id: {:?}", active.id);
+			}
 		};
 
+		// NOTE: currently we act as if this block has been finalized by BEEFY as we perform
+		// the validator set changes instantly (insecure). Once proper validator set changes
+		// are implemented this should be removed
+		self.best_finalized_block = Some(*notification.header.number());
```

Comment:
I think this is a little bit confusing and I think we still miss a bit of logic. Let me elaborate.

We have two block variables that both affect `should_vote_on`:
1. `best_finalized_block`
2. `best_block_voted_on`

I think we kept confusing the two and probably the implementation instructions were incorrect.

So the `should_vote_on` formula should be roughly something like that:
```
(best_observed_block - best_beefy_finalized_block).next_power_of_two()
```

The logic for current variables:
1. `best_finalized_block` = `best_observed_block`, but with extra `None` value right after starting up.
2. `best_block_voted_on` = in case the node is voting, this will get updated with last vote.

I think we need to change the logic (and maybe the name) of `best_block_voted_on` instead of `best_finalized_block`. Basically:
1. `best_finalized_block` should simply be updated every time we receive GRANDPA finality notification (IMHO we should even call it `best_grandpa_block`).
2. `best_beefy_block` should be updated in every case from this list:
    1. We conclude a round (i.e there exists `SignedCommitment`).
    2. We transition to new validator set.
   
The second case is obviously a temporary measure for `Milestone 1`.